### ISSUE TITLE
Conversation branching: continue a thread from an earlier point and reuse the shared prefix

### ISSUE THREAD

**Issue by fpt** [fpt/rs-gallium#284]
## The proposal

Let a client branch a conversation — continue from an earlier point in the
history rather than from its end — and have the shared prefix's KV cost nothing
the second time. The suggested surface was a `resume_from: <cache id>` on the
app-server, with the branch *tree* owned by the client (klein) and gallium
holding only "cache id → KV state, resume/extend". Motivating uses: ChatGPT-style
regenerate-from-turn-N, agent rollback ("go back and try the other tool call"),
and A/B trajectory generation for chosen/rejected training pairs.

Feasibility investigated below. Summary: **the inference half is already built
and needs no cache id; the missing half is the app-server, where it is small;
and the ambitious half — many live branches sharing one prefix — is blocked
upstream and should not be attempted yet.**

## 1. Reuse here is content-addressed, so a cache id would be a second, worse key

`LlamaLocalProvider::reserve` (`crates/gallium-agent/src/llm_local.rs:1913`)
picks a slot by scoring every idle one with `common_prefix_len` against the
**incoming token ids** and taking the maximum:

```rust
let warmest = pool.idle.iter().enumerate()
    .filter(|(_, s)| s.n_ctx >= needed)
    .map(|(i, s)| (common_prefix_len(&s.tokens, tokens), i))
    .max();
```

The candle backend does the same thing (`llm_candle.rs:466`, `reusable_prefix`).
Neither asks who the caller is or which conversation this is; both ask only what
the cache literally holds.

That is already the branching primitive. A branch's prompt *is* the parent's
prompt for its first N tokens, so a branch continuation matches the parent's slot
and evaluates only its own suffix, with no new concept. A `resume_from` id would
add a second key for the same lookup, one that can disagree with the tokens — and
`generate_reusing` (`:1683`) is deliberately built so that it cannot be lied to:
the comment there says matching is on ids "never on the assumption that the
prompt was appended to", because the next prompt is re-rendered through the jinja
template from scratch. An id-addressed restore reintroduces exactly the failure
that design excludes: a state put somewhere the token record does not describe,
which produces plausible wrong logits and which nothing downstream can detect.

So: **keep the address content-based.** Whatever is built here, the client sends
messages and gallium finds the cache.

## 2. The actual gap is that a thread cannot be seeded with a history

`handle_thread_start` builds a thread's messages from the profile preamble plus
`developerInstructions` and nothing else (`appserver/server.rs:1377-1388`); the
`Thread` response hard-codes `"turns": []` with the comment "Only ever populated
by the history methods gallium does not implement" (`:1475`). There is no
`thread/resume`, no fork, no way to say "this conversation begins here."

Branching from an earlier point means starting a thread whose history is the
parent's first N messages. There is no way to express that today, so a client
cannot branch at all, cache or no cache.

**And the existing workaround actively destroys the reuse this issue is about.**
klein already re-seeds history when a thread is replaced, through
`developerInstructions` — `../klein-cli/pkg/agentserver/runner.go:398`:

> developerInstructions ... is where a caller puts anything a fresh thread must
> begin with, **including the conversation so far**: the server holds a thread's
> history, so a thread started here has none.

That lands the entire prior conversation as one system message at the *front* of
the prompt (`server.rs:1387`). Two branches of one conversation then differ in
their first few hundred tokens instead of their last few, so `common_prefix_len`
returns roughly zero and every branch pays a full re-prefill. It is a
prompt-purity violation ([ADR 0001](docs/adr/0001-prompt-purity-and-explicit-context.md))
of the most expensive kind, and it is what a client has to do today.

Fixing this is the whole minimum-viable feature and it touches no cache code:
accept a prior message list at `thread/start`, place it as real messages in role
order, and the prefix match does the rest. The reuse then shows up in
`KvProvenance::SlotReuse` without anything else being written.

## 3. What a *shared-prefix tree* would cost, and why not yet

The endgame in the proposal — immutable shared prefix blocks plus branch-local
suffixes, vLLM-style — is a real thing llama.cpp can do, with two hard
qualifications.

**It requires `kv_unified`, which `llama-cpp-2` does not expose.**
`llama_kv_cache::seq_cp` is metadata-only — it just adds the destination seq id
to each cell's bitmask, no data copied — but *only when both sequences are in the
same stream* (`llama-kv-cache.cpp:461-487`). Streams are allocated
`n_stream = unified ? 1 : n_seq_max` (`:82`), `kv_unified` defaults to **false**
(`llama-context.cpp:3543`), and `llama-cpp-2` 0.1.156 exposes `n_seq_max`,
`n_rs_seq` and `swa_full` but **not `kv_unified`**. Without it, N branches means
N whole KV buffers and `seq_cp` becomes a cross-stream buffer copy — precisely
the "10k prefix × 10 copies" the proposal warns about, arrived at by accident.
The fix upstream is a one-field passthrough; until it lands there is no cheap
sharing to build on.

Also worth recording as a hazard: `llama_kv_cache::seq_cp` opens with
`if (other) { return; }` (`:449`, `TAG_KV_CACHE_SHARE_CELLS`) and returns void.
It is a silent no-op when the cache is linked to another context's memory.
Gallium never sets `ctx_other`, so this does not bite today, but any code built
on `seq_cp` must not assume it happened.

**Recurrent and hybrid models get no sharing at all.**
`llama_memory_recurrent::seq_cp` (`llama-memory-recurrent.cpp:235`) only aliases
the source's tail cell into the destination seq — there is no per-position log to
share, because that is the same property that makes partial rollback impossible
(the thing `Checkpoint` exists to work around). For LFM2 and the Qwen hybrids the
only branch primitive is the existing whole-state `Checkpoint`, which is
**~12 KiB per token of host memory** (`llm_local.rs`, `Checkpoint` docs). Ten
branches held at 10k tokens is ~1.2 GB of RAM, per model, and no part of it is
shared. DeepSeek-V4 cannot branch by state at all — `state_seq_get`/`set` does
not round-trip its cache and it is denylisted from checkpoints (#209).

So the tree pays off on pure-attention models (Gemma 4, GPT-OSS, Qwen3.8
dense) and is unavailable on exactly the models whose prefills hurt most.

**The candle backend is, unusually, the better-placed one.** Its caches are
gallium's own tensors: `KvCache::truncate` (`gallium-core/src/kv_cache.rs:174`)
rolls an attention layer back to any earlier position, and `CacheCheckpoint`
(`:266`) holds *only* the recurrent layers, because the attention half "needs
nothing kept". A genuine copy-on-write prefix tree there is a gallium-core
change with no upstream dependency and no `kv_unified` problem. That is worth
knowing, but it is a second implementation of a feature that has none yet, so it
is not where to start.

## 4. Relationship to #174

#174 (persistent tiered session state) already argues the structural half of
this, and states the prerequisite plainly: **a slot is not a session.** Branching
is a consumer of that refactor, not an independent feature — "ten conversations
behind two slots" and "ten branches behind two slots" are the same object with
the same hydration problem. Doing branching first would build a branch table on
top of a pool that still equates a session with a physical context.

The one thing branching adds to #174's model is that sessions form a *tree* with
shared prefixes rather than a flat set, which changes eviction (evicting a parent
that three branches share is not an LRU decision) but nothing else.

## Proposed scope

Staged, smallest first; each stage is useful alone.

1. **Seed a thread's history at `thread/start`.** A prior message list, placed as
   real messages in role order, replacing the `developerInstructions`
   history-stuffing klein does today. No cache code. This alone makes branching
   possible *and* makes it fast, because the content-addressed match already
   works. Wire shape should stay inside codex's protocol where it has one — see
   CLAUDE.md's rule about `skillCount`; an additive gallium-only field is how the
   first divergence always argues for itself.
2. **Raise the slot count for this case and say so.** `GALLIUM_KV_CACHE_SLOTS`
   defaults to 1, which means alternating between two branches evicts each
   other's tails every turn. Branching wants one slot per live branch and pays a
   whole KV cache for each — document it as the memory knob it is.
3. **Make the saving visible.** `KvProvenance` already distinguishes
   `SlotReuse` / `CheckpointRestore` / `CacheReset`; a branch turn should be
   measurable from a trace without instrumentation ("branched at token 8 412,
   evaluated 137").
4. **Then, only after #174's session/slot split lands**, consider a real prefix
   tree — and only for pure-attention models, and only once `kv_unified` is
   reachable from `llama-cpp-2`.

Explicitly **not** proposed: a `resume_from` cache id (§1), gallium holding
conversation-tree semantics (the proposal is right that this belongs in the
client), and any llama.cpp fork.

## Open questions

- Does a seeded history need a fingerprint of the model it was produced against?
  A branch re-seeded onto a *different* model is a correctness question about the
  transcript, not about the cache, and the cache handles itself by matching
  fewer tokens.
- `HarmonyProtocol`'s `Current date:` line is the known prompt-purity violation
  (CLAUDE.md, ADR 0001). It sits near the front of every GPT-OSS prompt, so a
  branch replayed the next day shares no prefix with its parent. Branching makes
  that cost concrete rather than theoretical.
- The A/B trajectory use (same prefix, different seed / reasoning effort /
  model) wants the *same* thread history with different sampling — which stage 1
  gives for free on one model, and which crosses providers in a way nothing here
  addresses. Worth its own issue if it turns out to be the primary use.

Related: #174 (the prerequisite refactor), #209 (DeepSeek-V4 state does not
round-trip), #170 (connection lifecycle vs. backend loading), #13.

**Comment by fpt** [comment]
## Correction to §3: `kv_unified` **is** reachable from `llama-cpp-2`, so stage 4 is not blocked upstream

Re-checked against the tree at 95aa42f. Every other claim in §3 verifies exactly;
this one does not.

`llama-cpp-2` 0.1.156 — the version `Cargo.lock` pins — exposes it:

```
llama-cpp-2-0.1.156/src/context/params/get_set.rs:713
    pub fn with_kv_unified(mut self, kv_unified: bool) -> Self
llama-cpp-2-0.1.156/src/context/params/get_set.rs:728
    pub fn kv_unified(&self) -> bool
```

0.1.154 has it too (7 hits in the same file), so this is not a new addition we
would be waiting on. The two other primitives a prefix tree needs are also
already bound:

```
llama-cpp-2-0.1.156/src/context/kv_cache.rs:137
    pub fn kv_cache_seq_cp(&mut self, seq_id_src, seq_id_dst, p0, p1)
llama-cpp-2-0.1.156/src/context/params/get_set.rs:113
    pub fn with_n_seq_max(mut self, n_seq_max: u32) -> Self
```

What is true is that gallium never sets any of them. `build_context`
(`llm_local.rs:2032`) builds every context with

```rust
LlamaContextParams::default().with_n_ctx(NonZeroU32::new(n)).with_n_batch(n)
```

so `n_seq_max` stays at its default and `kv_unified` at `false`
(`llama-context.cpp:3543`, as cited). Everything below that uses sequence 0 and
only sequence 0 — `clear_kv_cache_seq(Some(0), ..)` at `:1805` and `:1833`,
`state_seq_get(0, ..)` at `:2363`, `state_seq_set(&state, 0)` at `:2407`.

So the blocker in §3 is **ours, not upstream**: reaching cheap cross-branch
sharing is a `build_context` change (`.with_kv_unified(true).with_n_seq_max(k)`)
plus a slot that tracks more than one sequence id, not a `llama-cpp-2` PR. That
does not make stage 4 cheap — it still needs #174's session/slot split, and it is
still unavailable on the recurrent and hybrid models — but "wait for upstream" is
the wrong reason to defer it, and it is the reason most likely to be believed
without being rechecked.

Verified unchanged in the vendored 0.1.156 source, for the record:

- `n_stream(unified ? 1 : n_seq_max)` — `llama-kv-cache.cpp:82`. Confirms the
  cost argument: without `kv_unified`, N branches are N buffers and `seq_cp`
  takes the "cross-stream sequence copies require to copy the actual buffer
  data" path.
- `seq_cp` is metadata-only in-stream (`cells.seq_add(i, seq_id_dst)` under
  `if (s0 == s1)`), and opens with `if (other) { return; }`
  (`TAG_KV_CACHE_SHARE_CELLS`), returning void — the silent no-op hazard as
  described.
- `llama_memory_recurrent::seq_cp` aliases only the tail cell.
- `common_prefix_len` scoring in `reserve` (`:1913`), `reusable_prefix`
  (`llm_candle.rs:466`), `GALLIUM_KV_CACHE_SLOTS` default 1 (`:94`),
  `"turns": []` (`appserver/server.rs:1475`), thread messages seeded from
  preamble + `developerInstructions` only (`:1377-1388`), `KvProvenance`
  (`llm.rs:130`), `KvCache::truncate` (`gallium-core/src/kv_cache.rs:174`).

One line-number drift: `CacheCheckpoint` is `kv_cache.rs:275`, not `:266`.

Stages 1–3 are unaffected.
